Handle forks in the `branch-history` GitHub CI job

https://github.com/IntersectMBO/cardano-ledger/actions/runs/9843781035/job/27183341050?pr=4478

```
Run git fetch origin master amw/fix_allocation
  git fetch origin master amw/fix_allocation
  if git log --merges --oneline --decorate origin/master..origin/amw/fix_allocation | grep .; then
    echo 'Branch must not contain merges'
    false
  fi
  shell: /usr/bin/bash -e {0}
fatal: couldn't find remote ref amw/fix_allocation
Error: Process completed with exit code 128.
```

--- a/.github/workflows/haskell.yml
+++ b/.github/workflows/haskell.yml
@@ -346,13 +346,16 @@ jobs:
   branch-history:
     name: Check branch history
     runs-on: ubuntu-latest
-    if: ${{ github.base_ref != '' && github.head_ref != '' }}
+    if: ${{ github.base_ref != '' && github.ref != '' }}  # Only true for PRs
     steps:
     - uses: actions/checkout@v4
     - name: Ensure the branch doesn't contain any merges
       run: |
-        git fetch origin ${{ github.base_ref }} ${{ github.head_ref }}
-        if git log --merges --oneline --decorate origin/${{ github.base_ref }}..origin/${{ github.head_ref }} | grep .; then
+        PR_TARGET=${{ github.base_ref }}
+        PR_MERGE=${{ github.ref }}
+        PR_HEAD=${PR_MERGE%/merge}/head
+        git fetch origin -n --refmap= +$PR_TARGET:pr-target +$PR_HEAD:pr-head
+        if git log --merges --oneline --decorate pr-target..pr-head | grep .; then
           echo 'Branch must not contain merges'
           false
         fi

